### ISSUE TITLE
C# Code Style option for 'prefer var' should support `foreach` statements that use a different type than the returned enumerable

### ISSUE THREAD

**Issue by tannergooding** [dotnet/roslyn#14052]
**Issue**
Some generic types, such as `IEnumerable<out T>`, are covariant (this means you can use a more derived type than originally specified).

There are other generic types, such as `Action<T>`, which are contravariant (this means you can use a less derived type than originally specified).

The 'prefer var' C# Code Style option should allow the user to special case scenarios involving covariant or contravariant types.

**Repro**
- Create a new C# Console Application
- Modify the C# Code Style Options so that `var preferences: Elsewhere` is `prefer var` and has a severity of `Warning`
- Modify your `Program.cs` file to contain the following code:

``` C#
using System;
using System.Collections.Generic;

interface IContractV1
{
    string GetId();
}

interface IContractV2 : IContractV1
{
    string GetUniqueId();
}

class ContractFactory
{
    public IEnumerable<IContractV1> GetContracts()
    {
        return new List<IContractV2>();
    }
}

class Program
{
    static void Main(string[] args)
    {
        var contractFactory = new ContractFactory();

        foreach (IContractV2 contract in contractFactory.GetContracts())
        {
            Console.WriteLine(contract.GetUniqueId());
        }
    }
}
```

Note that `IContractV2` in `foreach(IContractV2 contract in contractFactory.GetContracts())` has a warning diagnostic.

**Workarounds**
A user can use LINQ to perform the explicit cast on each interface member:

``` C#
foreach (var contract in contractFactory.GetContracts().Select((c) => (IContractV2)(c)))
{
    Console.WriteLine(contract.GetUniqueId());
}
```

A user can also create a temporary variable scoped to the foreach loop and explicitly perform the cast:

``` C#
foreach (var contractV1 in contractFactory.GetContracts())
{
    var contractV2 = (IContractV2)(contractV1);
    Console.WriteLine(contractV2.GetUniqueId());
}
```

**Comment by CyrusNajmabadi** [comment]
Note: this is not really related to covariance.  This is how C# will insert a cast for you based on the type of your foreach variable.  in this case we need to recognize that the implicit cast is signiicant, and not offer to simplify to 'var'.

**Comment by jmarolf** [comment]
fixed with #17102
